common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -428,121 +450,138 @@ public boolean close(T trackedObject) {
-                   .append(NEWLINE);
+            if (duped > 0) {
+                buf.append(": ")
+                        .append(dropped)
+                        .append(" leak records were discarded because they were duplicates")
+                        .append(NEWLINE);
             }
+
             if (dropped > 0) {
-                buf.append(dropped)
-                   .append(" leak records were not sampled because the leak record sample count is limited to ")
-                   .append(MAX_SAMPLED_RECORDS)
+                buf.append(": ")
+                   .append(dropped)
+                   .append(" leak records were discarded because the leak record count is targeted to ")
+                   .append(TARGET_RECORDS)
                    .append(". Use system property ")
-                   .append(PROP_MAX_SAMPLED_RECORDS)
+                   .append(PROP_TARGET_RECORDS)
                    .append(" to increase the limit.")
                    .append(NEWLINE);
             }
 
-            int records = idx - removed;
-            buf.append("Recent access records: ").append(records).append(NEWLINE);
-
-            if (records > 0) {
-                // The array may not be completely filled so we need to take this into account.
-                for (int i = records - 1; i >= 0; i --) {
-                    buf.append('#')
-                       .append(i + 1)
-                       .append(':')
-                       .append(NEWLINE)
-                       .append(array[i]);
-                }
-            }
-
-            buf.append("Created at:")
-               .append(NEWLINE)
-               .append(creationRecord);
-
             buf.setLength(buf.length() - NEWLINE.length());
             return buf.toString();
         }
     }
 
-    private static final class Record extends Throwable {
+    private static final AtomicReference<String[]> excludedMethods = new AtomicReference<String[]>(new String[0]);
 
-        private static final Set<String> STACK_TRACE_ELEMENT_EXCLUSIONS = new HashSet<String>();
-        static {
-            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.util.ReferenceCountUtil.touch");
-            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.buffer.AdvancedLeakAwareByteBuf.touch");
-            STACK_TRACE_ELEMENT_EXCLUSIONS.add("io.netty.buffer.AbstractByteBufAllocator.toLeakAwareBuffer");
-            STACK_TRACE_ELEMENT_EXCLUSIONS.add(
-                    "io.netty.buffer.AdvancedLeakAwareByteBuf.recordLeakNonRefCountingOperation");
+    public static void addExclusions(Class clz, String ... methodNames) {
+        Set<String> nameSet = new HashSet<String>(Arrays.asList(methodNames));
+        // Use loop rather than lookup.  This avoids knowing the parameters, and doesn't have to handle
+        // NoSuchMethodException.
+        for (Method method : clz.getDeclaredMethods()) {
+            if (nameSet.remove(method.getName())) {
+                if (nameSet.isEmpty()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1066)

Reply:
+1

Reply:
done